### ISSUE TITLE
Ability to remove rules by name

### ISSUE THREAD

**Issue by alhi3** [CacheControl/json-rules-engine#156]
I'm building a service using this engine. The users are able to add and remove rules whenever they want. I'm having troubles on the removal process.

Every time a user adds a rule, I create a new rule using `let  rule = new engine.Rule(options)`. With that rule I do two things:

1.- Save the JSON to the filesystem, so every time I need to reload the service I could recreate the rules.
2.- Add the rule to then engine.

This works great. The problem that I have is when I try to remove a rule. For that process the user gives me the name of the rule an I do the following:

1.- Find in the filesystem the `.json` file that correspond to the name the user gives me.
2.- Create a json object from that file:
`let ruleFromJson = JSON.parse(fs.readFileSync(file, { encoding: 'UTF-8' }))`
3.- Create a Rule from that object:
`let rule = new engine.Rule(ruleFromJson);`
4.- Remove the rule:
`engine.removeRule(rule);`

After doing that if I run the engine passing a fact that match the deleted rule, the engine still throw `onSuccess` with that rule.

What I'm doing wrong?

Versions:
- Node: 10.16.0
- JSON Rules Engine: 4.0.0

**Comment by CacheControl** [comment]
@alopez-hi What does `removeRule()` return? If the return value is `false`, it means the rule wasn't found in the engine and therefor couldn't be removed.

I strongly suspect this is the case, and the reason the rule can't be found is because the object that is being re-hydrated from the file system is not the same reference as the object currently in the engine.

This is a missing feature; what you're trying to do should be possible, we just don't support it yet. 

## Quickfix

As a quick fix, you can iterate over your `engine.rules` property, and manually `delete` the appropriate element of the array, e.g.

```js
engine.rules.forEach((rule, idx) => {
  // use object equality library of choice, here I'm using lodash
  // https://lodash.com/docs/4.17.11#isEqual
  if (_.isEqual(rule, ruleToDelete)) {
    delete engine.rules[idx]
    engine.prioritizedRules = null // clear the internal rules cache
  }
})
```

## Long term feature request

Ability to `removeRule()` using rule.name:

```js
const ruleName = 'my-rule'
new Rule({ name: ruleName, conditions, event })

engine.removeRule(ruleName)
```

**Comment by alopez-hi** [comment]
@CacheControl , you're right. The engine method `removeRule()` return false. As you do in the quick fix I manually remove the rule. We'll wait until this feature to come out. Thanks!

**Comment by ilanskogan** [comment]
@CacheControl Hi, this feature still isn't available?

**Comment by CacheControl** [comment]
@ilanskogan that's correct; it has not been implemented.

**Comment by ilanskogan** [comment]
@CacheControl from what I see all it takes is to change in the engine the this.rules to object instead of an array and then in each place you have reference to this.rules you need to replace it with Object.values(this.rules)
this will allow me to update/remove rule by a key (which is id) .
am I missing something? i guess if it was so simple someone already would make a PR for this

**Comment by ilanskogan** [comment]
@CacheControl if i am correct i would like to submit a pr, I have first version of it need to test it a little more

**Comment by CacheControl** [comment]
@ilanskogan I agree this should be an easy thing to implement; I'm not sure why no one has chosen to try it. I'll happily accept a PR.

**Comment by ilanskogan** [comment]
@CacheControl alright i will work on one.
Should i keep the current interface or i can break it?
For example i would like an option to delete rule by key(rule_id) should the "removeRule" still need to support Rule instance as parameter and i need to add removeRuleById function or i can consider this as major release and therefore breaking changes are fine here?

**Comment by ilanskogan** [comment]
@CacheControl I made a PR for this. please take a look [https://github.com/CacheControl/json-rules-engine/pull/264](url)

**Comment by eran-peled** [comment]
It will be epic, if we will have @ilanskogan functionallity.

**Comment by CacheControl** [comment]
Now available in **6.1.X**
